### ISSUE TITLE
residential_quota_downselect exits with empty downselect

### ISSUE THREAD

**Issue by lixiliu** [NREL/buildstockbatch#391]
**Describe the bug**
Ran `buildstock_local --samplingonly` with resample=true and found that sampler can exit when downselected buildstock.csv turns up empty.

I thought with resample=True, sampler will iteratively try higher sampling count until downselected buildstock has >= n_datapoints specified.

Can we allow sampler to be run with a different random seed?
```
Traceback (most recent call last):
  File "/Users/lliu2/.local/bin/buildstock_local", line 33, in <module>
    sys.exit(load_entry_point('buildstockbatch', 'console_scripts', 'buildstock_local')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lliu2/Documents/GitHub/buildstockbatch/buildstockbatch/utils.py", line 107, in run_with_error_capture
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lliu2/Documents/GitHub/buildstockbatch/buildstockbatch/local.py", line 364, in main
    batch.run_batch(n_jobs=args.j, measures_only=args.measures_only, sampling_only=args.samplingonly)
  File "/Users/lliu2/Documents/GitHub/buildstockbatch/buildstockbatch/local.py", line 218, in run_batch
    buildstock_csv_filename = self.sampler.run_sampling()
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lliu2/Documents/GitHub/buildstockbatch/buildstockbatch/sampler/downselect.py", line 126, in run_sampling
    raise RuntimeError('There are no buildings left after the down select!')
RuntimeError: There are no buildings left after the down select!
```

**To Reproduce**
 yaml used on EUSS v1 ResStock: *.yml converted to .txt because GitHub does not support .yml file upload.
[national_baseline.txt](https://github.com/NREL/buildstockbatch/files/12588145/national_baseline.txt)

**Expected behavior**
Returned downselected buildstock of size >= 1.

**Logs**

[buildstockbatch_crash_details.log](https://github.com/NREL/buildstockbatch/files/12588160/buildstockbatch_crash_details.log)

**Comment by rajeee** [comment]
It doesn't iteratively try higher number of samples. It does 350K sample first, and determines the expected number of samples needed to get the required data points. In your case (based on the crash log), 350K sampling resulted in 125 rows after downsampling. So, it determined that 350K/125 = 2800 sampling would result in 1 datapoints after downselect. However, due to how quota sampling works, it didn't get any datapoint after downselect - hence the error.

It's not really a bug - more of a limitation of the downselect sampler. You don't usually get exactly the requested number of datapoints after downselect. Typically, it's fine (you request 10K and you get 9998 samples), but in this case you requested 1 and got zero - hence the crash.

If you can request at least a couple, then you should be fine. If you strictly need only one, you can always remove the rows later.

**Comment by lixiliu** [comment]
I see, thanks for the explanation. I am surprised it doesn't then try 3000 sampling to see if it arrives at something. I understand it wasn't how the sampler was designed to do.

Related to what I am doing, there's a desire based on a few recent data requests to get the most "typical" or "representative" single building for a given set of criteria. I'd argue that downselected sampling of 1 comes close to providing that but not always because the selection of attributes is stochastic. In the case where via downselected sampling, there are 10 samples for the criteria, we have no idea which building is the most representative and therefore know which to drop. There's a mathematically way to figure this out with joint prob but it's not really doable.

The math way: the best way to determine this (and to determine most typical features without sampling) is the through a downselected joint probability of all housing characteristics (HC) but my understanding is it's impossible to compute given the size of the HC network. As with sampling_probability calculations in resstock-estimation, we can only compute and retain the joint prob of a handful of HCs at a time. But if we allow these joint probs to be filtered down early, we may be able to do more. IDK, just my thoughts.

**Comment by shorowit** [comment]
Related: https://github.com/NREL/buildstockbatch/issues/319

**Comment by afontani** [comment]
Closed because #319 is the preferred path.
